### DISCUSSION TITLE
How to expose background image for applications?

### DISCUSSION THREAD

**Discussion by ghost** [wezterm/wezterm#1537] [Q&A]
I'm building in opacity into my windowing system (see below), and I would like to be able to expose the background image the user sets in wezterm while also doing my own alpha-blending over it.  Is that possible?

So imagine a scenario like this, but if I had a clear background (no hatch) drawn first with default colors (SGR 39/49) and then overlaid with {space} + new background color.  It would be awesome if the user's background image/animation could still show through.  Basically I'm asking if there is a way to set an alpha value in a color selection, even if it's wezterm-specific for now.  My hunch is not, but then I haven't been able to alpha blend until yesterday so figured I would ask.

![xterm_translucent](https://user-images.githubusercontent.com/4357501/149151153-52ea14e5-6c6a-4991-83bf-52226e78bcd9.png)

**Comment by wez** [comment]
I'm open to exploring this!

We could extend the true color and palette modification escape sequences to support specifying alpha channel values and, while we're in there thinking about that, also consider specifying how to use 10 bits per channel, which is something @dankamongmen and I briefly chatted about in the past and which I think we've both tentatively made room for in our respective implementations.

![image](https://user-images.githubusercontent.com/117777/149175294-f62043f9-159f-4d34-b018-e2d43eba7875.png)
there are a few different layers to cut across, and those have their own set of color related types to wrangle!

termwiz's internal color type currently supports 8 or 10 bits per channel with no alpha:

https://github.com/wez/wezterm/blob/f6e56aab10b51fef3488f25b16d7a9ca6f74b09b/termwiz/src/color.rs#L89-L92

Adding an alpha channel will mean changing something, looking at the code right now we already have small vs heavyweight color types:

https://github.com/wez/wezterm/blob/f6e56aab10b51fef3488f25b16d7a9ca6f74b09b/termwiz/src/cell.rs#L14-L17
https://github.com/wez/wezterm/blob/f6e56aab10b51fef3488f25b16d7a9ca6f74b09b/termwiz/src/cell.rs#L76-L85

So perhaps we can "just" add a `TrueColorWithAlpha(RgbColor, f32)` variant here:

https://github.com/wez/wezterm/blob/f6e56aab10b51fef3488f25b16d7a9ca6f74b09b/termwiz/src/color.rs#L404-L410

and similarly for `ColorAttribute` over here:

https://github.com/wez/wezterm/blob/f6e56aab10b51fef3488f25b16d7a9ca6f74b09b/termwiz/src/color.rs#L430-L440

(those `Falback` variants are used by the termwiz library when it believes that the terminal doesn't support true color.  They're
not meaningful in the context of the `term` subdirectory, but since those types are shared between them, we have to consider
them both together)

The actual terminal palette type is over here; I think we could probably redefine this to be an array of [RgbaTuple](https://github.com/wez/wezterm/blob/f6e56aab10b51fef3488f25b16d7a9ca6f74b09b/termwiz/src/color.rs#L56); the space requirements are less of a concern because the palette is low cardinality and is already a lookup table for simple indices:

https://github.com/wez/wezterm/blob/f6e56aab10b51fef3488f25b16d7a9ca6f74b09b/term/src/color.rs#L9-L10

(caveat: `RgbaTuple` is composed of linear (rather than SRGB) float values. Having them be linear is actually a plus because we can then avoid converting them in the gui layer later on.  However, in Rust, floats are not `Eq` but are `PartialEq`. The `Palette256` type is currently `Eq`. We might be able to remove `Eq` from it without a headache)

The compiler will help you find the fanout... ultimately these things bubble up to the gui over here:

https://github.com/wez/wezterm/blob/f6e56aab10b51fef3488f25b16d7a9ca6f74b09b/wezterm-gui/src/termwindow/render.rs#L2745-L2752

which is where we take the (currently alpha-less) colors resolved from the terminal palette, linearize it from SRGB and add an alpha channel value and return this gui level type:

https://github.com/wez/wezterm/blob/f6e56aab10b51fef3488f25b16d7a9ca6f74b09b/color-types/src/lib.rs#L207-L209

If we're tracking alpha all the way through, together with pre-linearized color channels, those functions become simpler: I think they just need to multiply the `alpha` parameter (which comes from the `window_background_opacity` or `text_background_opacity` configuration) by the alpha value in the resolved color but pass the other values through as-is.

**Comment by ghost** [comment]
My head is going to hurt in a bit. :-)  It's lovely, _thank you_ so much for pulling this out.

So I'm imaging a rendering pass like this:
1. Blit X11/Wayland/etc. background image (assuming we have that kind of transparency -- sometimes that is left to the windowing system, sometimes programs peek at the desktop underneath).
2. Blit wezterm background image+alpha.
3. Blit undertext image layer(s) (kitty, future Good Image Protocol).
4. Blit cell background+alpha.  I'm assuming we would extend SGR 38/48, so could have an alpha on each layer.  But it may be better to have a fore+back single alpha, and then we don't hook to SGR 38/48 but to something else.
5. Blit cell foreground glyphs+alpha.  Glyphs could be invisible here too (e.g. blinking cycle, or actual ' ') -- so for that case we have an effective NOP (every pixel would be transparent).
6. Blit overtext image layer(s) (kitty, sixel, iTerm2, future Good Image Protocol).

...let me think more.  Maybe in a bit I can try my hand at a PR...

**Reply by wez** [reply]
> My head is going to hurt in a bit. :-) It's lovely, _thank you_ so much for pulling this out.

You're welcome!

> So I'm imaging a rendering pass like this:

I've stuck some pointers the relevant bits of this below, however, I think what you're asking for is essentially:

* Add alpha to the color attributes/palettes.  It defaults to 100%
* Define an escape sequence/extension to the existing escapes that modify the color palette to support alpha
* Define an escape sequence/extension to the existing escapes that set a true color attribute to make them support alpha

With those in place, I *think* wezterm will pass things through and honor them.
If that is true, then hopefully you won't need to worry too much about the below!

> 1. Blit X11/Wayland/etc. background image (assuming we have that kind of transparency -- sometimes that is left to the windowing system, sometimes programs peek at the desktop underneath).

That pass doesn't apply to wezterm: window transparency is left to the windowing environment: we just set the alpha value in our framebuffer and let the environment composite if it is capable.

> 2. Blit wezterm background image+alpha.

That starts here; the logic is blending into the per-pane render routine partly for legacy reasons, but it allows us to reason about whether the pane background should actually bleed out to the whole window if there is just one pane with a redefined background color:

https://github.com/wez/wezterm/blob/f6e56aab10b51fef3488f25b16d7a9ca6f74b09b/wezterm-gui/src/termwindow/render.rs#L1034-L1036

We divide rendering into three passes/layers:

* 0 - Background (color, image, gradient) which is drawn with regular alpha blending
* 1 - Glyphs which are drawn with dual source blending for hardware assisted subpixel AA
* 2 - Underlines/strikethrough which is drawn with regular alpha blending

the render logic makes a single pass through the panes, accumulating quads into those layer categories, then at the end,
once all textures and quads have been successfully allocated (we'll reset, grow and retry if we need more), we submit
an opengl draw call for each layer.

> 3. Blit undertext image layer(s) (kitty, future Good Image Protocol).

That happens here:
https://github.com/wez/wezterm/blob/f6e56aab10b51fef3488f25b16d7a9ca6f74b09b/wezterm-gui/src/termwindow/render.rs#L2024-L2036

> 4. Blit cell background+alpha.  I'm assuming we would extend SGR 38/48, so could have an alpha on each layer.  But it may be better to have a fore+back single alpha, and then we don't hook to SGR 38/48 but to something else.

https://github.com/wez/wezterm/blob/f6e56aab10b51fef3488f25b16d7a9ca6f74b09b/wezterm-gui/src/termwindow/render.rs#L1846-L1849

> 5. Blit cell foreground glyphs+alpha.  Glyphs could be invisible here too (e.g. blinking cycle, or actual ' ') -- so for that case we have an effective NOP (every pixel would be transparent).

The heavy lifting to figure out the fg/bg/blink etc is triggered here:
https://github.com/wez/wezterm/blob/f6e56aab10b51fef3488f25b16d7a9ca6f74b09b/wezterm-gui/src/termwindow/render.rs#L1914-L1919

This is the bit where we just skip invisible glyphs; not doing so results in ghostly AA artifacts even though the fg and bg are the same:
https://github.com/wez/wezterm/blob/f6e56aab10b51fef3488f25b16d7a9ca6f74b09b/wezterm-gui/src/termwindow/render.rs#L2084

> 6. Blit overtext image layer(s) (kitty, sixel, iTerm2, future Good Image Protocol).

https://github.com/wez/wezterm/blob/f6e56aab10b51fef3488f25b16d7a9ca6f74b09b/wezterm-gui/src/termwindow/render.rs#L2167-L2171

> ...let me think more. Maybe in a bit I can try my hand at a PR...

Eagerly anticipating this :-)

**Comment by ghost** [comment]
How I'm doing it in the TUI side: https://gitlab.com/klamonte/jexer/-/issues/88 (Plus mucho credit to notcurses for the inspiration and the key clue. :) )

I'm trying to figure out how to make something like composing translucent windows very easy on the app side, and also being able to punch through to the background.  (This would be easiest from the TUI side with a real Z axis even for text, e.g. [the cube](https://gitlab.com/klamonte/jexer/-/blob/master/docs/images.md).  Then you could set alpha on an entire layer, set layer to draw on, and away you go.  But no one will implement that.)

Hmm, maybe I'm being too dense.  I'm blasting 100k+ per frame sometimes and things are fine and performant.  A simple "set alpha" command might be all that is needed, and something like jexer / ncurses / notcurses could just suck it up and redraw all layers with correct alpha for every damaged cell on the topmost layer if we want the background to bleed through.

**Reply by wez** [reply]
Scoping the problem to just a single layer and getting the terminal background to shine through does feel as simple as setting the alpha channel appropriately.

More complex layering feels like it would benefit from the TE doing the compositing, and it would need to support proper layers for that.
I need to read your cube doc properly (literally just skimmed it real quick so far): looks really interesting at first glance, and I'm interested in the idea of supporting layers: that was something that crossed my mind while writing up my initial response above.

**Comment by AutumnMeowMeow** [comment]
(Oh whew! I thought GH had killed this whole thread. Yay, glad your outline is still here. :) )

**Reply by wez** [reply]
I was looking for this the other day, but failed to find it, when I was working on 240026de48f214f8ca11ac4816f256f3d3b00326 and 8cb74c62d203303e6de823c935bbe0340314542e
Those commits allow the text selection colors to include an alpha value, which tackles some of the boring bits around having an appropriate type for the color+alpha, but not all of what I mentioned above
